src/vs/workbench/api/browser/mainThreadEditorTabs.ts

```diff
@@ -27,31 +30,125 @@ export class MainThreadEditorTabs {
 		}
-
 		this._proxy.$acceptEditorTabs(tabs);
 	}
+
+	private _onDidTabOpen(event: IEditorsChangeEvent): void {
+		if (event.kind !== GroupChangeKind.EDITOR_OPEN || !event.editor) {
+			return;
+		}
+		if (!this._tabModel.has(event.groupId)) {
+			this._tabModel.set(event.groupId, []);
+		}
+		const editor = event.editor;
+		const tab = {
+			viewColumn: editorGroupToColumn(this._editorGroupsService, event.groupId),
+			label: editor.getName(),
+			resource: editor instanceof SideBySideEditorInput ? EditorResourceAccessor.getCanonicalUri(editor, { supportSideBySide: SideBySideEditor.PRIMARY }) : EditorResourceAccessor.getCanonicalUri(editor),
+			editorId: editor.editorId,
+			isActive: (this._editorGroupsService.activeGroup.id === event.groupId) && this._editorGroupsService.activeGroup.isActive(editor)
+		};
+		this._tabModel.get(event.groupId)?.push(tab);
+		// Update the currently active tab which may or may not be the opened one
+		if (tab.isActive) {
+			if (this._currentlyActiveTab) {
+				this._currentlyActiveTab.tab.isActive = (this._editorGroupsService.activeGroup.id === this._currentlyActiveTab.groupId) && this._editorGroupsService.activeGroup.isActive({ resource: URI.revive(this._currentlyActiveTab.tab.resource), options: { override: this._currentlyActiveTab.tab.editorId } });
+			}
+			this._currentlyActiveTab = { groupId: event.groupId, tab };
+		}
+	}
+
+	private _onDidTabClose(event: IEditorsChangeEvent): void {
+		if (event.kind !== GroupChangeKind.EDITOR_CLOSE || event.editorIndex === undefined) {
+			return;
+		}
+		this._tabModel.get(event.groupId)?.splice(event.editorIndex, 1);
+		this._findAndUpdateActiveTab();
+
+		// Remove any empty groups
+		if (this._tabModel.get(event.groupId)?.length === 0) {
+			this._tabModel.delete(event.groupId);
+		}
+	}
+
+	private _onDidGroupActivate(event: IEditorsChangeEvent): void {
+		if (event.kind !== GroupChangeKind.GROUP_INDEX) {
+			return;
+		}
+		this._findAndUpdateActiveTab();
+	}
+
+	private _findAndUpdateActiveTab() {
+		// Go to the active group and update the active tab
+		const activeGroupId = this._editorGroupsService.activeGroup.id;
+		this._tabModel.get(activeGroupId)?.forEach(t => {
+			if (t.resource) {
+				t.isActive = this._editorGroupsService.activeGroup.isActive({ resource: URI.revive(t.resource), options: { override: t.editorId } });
+			}
+			if (t.isActive) {
+				if (this._currentlyActiveTab) {
+					this._currentlyActiveTab.tab.isActive = (this._editorGroupsService.activeGroup.id === this._currentlyActiveTab.groupId) && this._editorGroupsService.activeGroup.isActive({ resource: URI.revive(this._currentlyActiveTab.tab.resource), options: { override: this._currentlyActiveTab.tab.editorId } });
+				}
+				this._currentlyActiveTab = { groupId: activeGroupId, tab: t };
+				return;
+			}
+		}, this);
+	}
+
+	private _updateTabsModel(event: IEditorsChangeEvent): void {
+		// Call the correct function for the change type
+		switch (event.kind) {
```

Comment:
Here for example, if we had an array of events and one of the events is `GROUP_INDEX` you wouldn't have to bother to update individual tabs because you recreate the tabs model anyway.

Reply:
I went ahead and implemented this. Not sure if I'm a fan though because now I use a labelled break which feels like a `goto` and a bad programming paradigm. Let me know what you think of the implementation.

Reply:
Actually, all tests fail now. Do you have any ideas as I didn't really change anything big just the way events are processed but I can't seem to figure out why it now reports everything regarding tabs wrong. It works when debugging  but when running the script without a debugger attached produces the wrong results. Leads me to believe we won't be able to go ahead with this.